IPs/RSKIP178.md

```diff
@@ -0,0 +1,375 @@
+
+The block in encoded as:
+
+```
+blockHeader, transactions, uncles
+```
+
+And *uncles* is a list of RSK block headers. We define a **heavyblock** to be a *mergeMiningProof* with certain properties that will be verified in consensus.
+
+We extend the list uncles to include, after all RSK block headers, a number of heavyblock elements (defined later). To maintain semantic coherence, the *uncles* field is renamed *difficultyContributors*. The field *uncleCount* is renamed as *difficultyContribution*.
+
+*difficultyContributors* will be an RLP list two two elements: the first is the old *uncles* list. The next is the new *heavyblocks* list.
+
+```
+difficultyContributors  = { uncles  , heavyblocks }
+```
+
+*difficultyContributors* will also be an RLP list of two elements: the first, called *immediateContribution*, represents the difficulty contributed by uncles. The second, called *delayedContribution*, represents the contribution made by confirmation heavyblocks.
+
+```
+difficultyContributors = { immediateContribution , delayedContribution }
+```
+
+The maximum number of uncles referenced in *uncles* will be reduced from 10 to 6.
+
+The maximum number of heavyblocks referenced in *heavyblocks* will be 2.
+
+A Bitcoin block header contains these fields:
+
+| Field          | Purpose                                                      | Size (Bytes) |
+| -------------- | ------------------------------------------------------------ | ------------ |
+| Version        | Block version number                                         | 4            |
+| hashPrevBlock  | 256-bit hash of the previous block header                    | 32           |
+| hashMerkleRoot | 256-bit hash based on all of the transactions in the block   | 32           |
+| Time           | Current [block timestamp](https://en.bitcoin.it/wiki/Block_timestamp) as seconds since 1970-01-01T00:00 UTC | 4            |
+| Bits           | Current [target](https://en.bitcoin.it/wiki/Target) in compact format | 4            |
+| Nonce          | 32-bit number                                                | 4            |
+
+We define how to convert the Bits field into a target difficulty that can be compared with the RSK block difficulty.
+
+The Bits field is stored in a compact format. Let Bits be a byte array *{ E,Q1,Q2,Q3 }*. Let M be the integer value *(Q1 << 16 + Q2 << 8 + Q3)* (that is the *Q* array interpreted in big-endian). The Bitcoin target is computed as *M\*2^(8(E-3))*. We define the *btcDifficulty0 = 2^256 / target*. Note that the *btcDifficulty0* is *2^32* times higher than the Bitcoin block difficulty.
+
+### Validation
+
+We'll define the function *getPoWInfo()* to check the validity of all kinds of blocks (mainchain, uncles and elements in the *heavyblocks* list). This means that the function that checks the validity of the proof of work of normal RSK blocks is changed to call getPoWInfo(). This is because some of the RSK blocks can be anchor heavyblocks.
```

Comment:
We define the function *getPoWInfo()* to check the validity of all kinds of blocks (mainchain, uncles and elements in the *heavyblocks* list). This means that the function that checks the validity of the proof of work of normal RSK blocks is changed to call getPoWInfo(), since some of the RSK blocks can be anchor heavyblocks.